### PR TITLE
msteele/APPEALS-8878 Update RSPEC for SendNotificationJob

### PR DESCRIPTION
**Pull request by msteele96**
Resolves APPEALS-8878, subtask APPEALS-8884

### Description
Added RSPEC for APPEALS 8878

### CODE REVIEW

File: spec/jobs/send_notification_job_spec.rb [hunk 1]
```diff
@@ -100,30 +104,30 @@
...
+      #   }
+      #   allow_any_instance_of(SendNotificationJob)
+      #     .to receive(:audit_params) { audit_creation_params }
+      #   allow_any_instance_of(SendNotificationJob)
+      #     .to receive(:audit_params)
+      #     .with(message, good_response, notification_events_id, notification_type)
+      #   perform_enqueued_jobs do
+      #     SendNotificationJob.perform_later(message)
+      #   end
+      # end
```
Comment (jshields13) [inline spec/jobs/send_notification_job_spec.rb:130]: Why is this commented out?

File: spec/jobs/send_notification_job_spec.rb [hunk 2]
```diff
@@ -177,20 +181,20 @@
...
+      #       bad_participant_id,
+      #       appeal_id,
+      #       email_template_id,
+      #       appeal_status
+      #     )
+      #   expect(Rails.logger).to receive(:error).with(/Failed with error:/)
+      #   perform_enqueued_jobs do
+      #     SendNotificationJob.perform_later(message)
+      #   end
+      # end
```
Comment (jshields13) [inline spec/jobs/send_notification_job_spec.rb:197]: Why is this commented out?

File: spec/jobs/send_notification_job_spec.rb (hunk 1 above)
Comment (msteele96) [reply spec/jobs/send_notification_job_spec.rb:130]: audit_params is no longer in the file, I was keeping it around until I could figure out if the test could be reused and altered slightly to test for something new

File: spec/jobs/send_notification_job_spec.rb (hunk 2 above)
Comment (msteele96) [reply spec/jobs/send_notification_job_spec.rb:197]: VaNotifyService.send_notifications is no longer called. Same reasoning as above

File: app/jobs/send_notification_job.rb [hunk 3]
```diff
@@ -91,14 +99,25 @@ def log_error(error_message)
...
+    # if FeatureToggle.enabled?(:va_notify_email)
+    #   notification_type = "Email"
+    # elsif FeatureToggle.enabled?(:va_notify_email, :va_notify_sms)
+    #   notification_type = "Email and SMS"
+    # elsif FeatureToggle.enabled?(:va_notify_sms)
+    #   notification_type = "SMS"
+    # else
+    #   notification_type = "None"
+    # end
+
```
Comment (jshields13) [inline app/jobs/send_notification_job.rb:120]: Please remove commented out code

File: app/jobs/vanotify/fetch_all_active_ama_appeals_job.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        appeal_type: "Appeal",
+        status: %w[assigned on hold in progress],
+        closed_at: nil
+      ).uniq(&:appeal_id)
+      active_ama_appeals_tasks.each do |task|
+        active_ama_appeals.push(task.ama_appeal)
+      end
+      active_ama_appeals
+    end
+end
```
Comment (jshields13) [inline app/jobs/vanotify/fetch_all_active_ama_appeals_job.rb:25]: Why is this in your PR?

File: app/jobs/vanotify/fetch_all_active_legacy_appeals_job.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      appeal_type: "LegacyAppeal",
+      status: %w[assigned on hold in progress],
+      closed_at: nil
+    )
+    active_legacy_appeals_root_tasks.each do |task|
+      active_legacy_appeals.push(task.appeal)
+    end
+    active_legacy_appeals
+  end
+end
```
Comment (jshields13) [inline app/jobs/vanotify/fetch_all_active_legacy_appeals_job.rb:25]: Why is this in your PR?

File: app/models/prepend/va_notify/appellant_notification.rb [hunk 6]
```diff
@@ -26,73 +26,44 @@ class NoAppealError < StandardError; end
...
+    message_attributes
   end
 
   def self.notify_appellant(
     appeal,
     template_name
   )
     msg_bdy = create_payload(appeal, template_name)
-    SendNotificationJob.perform_later(msg_bdy)
+    SendNotificationJob.perform_later(msg_bdy.to_json)
```
Comment (jshields13) [inline app/models/prepend/va_notify/appellant_notification.rb:62]: Is there nay possibility of .to_json throwing an error?

File: app/services/external_api/va_notify_service.rb [hunk 7]
```diff
@@ -18,13 +19,20 @@ class ExternalApi::VANotifyService
...
-    def send_notifications(participant_id, appeal_id, email_template_id, status = "")
-      email_response = send_va_notify_request(email_request(participant_id, appeal_id, email_template_id, status))
-      Rails.logger.info(email_response)
+    def send_email_notifications(participant_id, notification_id, email_template_id, status = "")
+      email_response = send_va_notify_request(email_request(participant_id, notification_id, email_template_id, status))
+      log_info(email_response)
       email_response
     end
+    
+      # Send the sms notifications
```
Comment (jshields13) [inline app/services/external_api/va_notify_service.rb:30]: Please add comments like tomdoc for ruby

Good example is comments for method on line 24

File: app/services/external_api/va_notify_service.rb [hunk 8]
```diff
@@ -70,11 +78,11 @@ def base64url(source)
     end
 
     # Build an email request object
```
Comment (jshields13) [inline app/services/external_api/va_notify_service.rb:80]: More verbose comments please

File: app/services/external_api/va_notify_service.rb [hunk 9]
```diff
@@ -92,15 +100,16 @@ def email_request(participant_id, appeal_id, email_template_id, status)
     end
 
     # Build a sms request object
-    def sms_request(participant_id, appeal_id, sms_template_id, status)
+    def sms_request(participant_id, notification_id, sms_template_id, status)
```
Comment (jshields13) [inline app/services/external_api/va_notify_service.rb:103]: More verbose comments please

File: app/services/external_api/va_notify_service.rb [hunk 10]
```diff
@@ -146,5 +155,9 @@ def send_va_notify_request(query: {}, headers: {}, endpoint:, method: :get, body
       end
     end
     # rubocop:enable Metrics/MethodLength
+
+    def log_info(info_message)
```
Comment (jshields13) [inline app/services/external_api/va_notify_service.rb:159]: More verbose comments please

**Review by jshields13 — COMMENTED** [review]
All in all very impressed with this code and the attention to detail in the unit test is amazing. I just have a few comments/questions then will be ready to merge PR

File: app/models/prepend/va_notify/appellant_notification.rb (hunk 6 above)
Comment (msteele96) [reply app/models/prepend/va_notify/appellant_notification.rb:62]: .to_json works on nil and will work on the object when it's created. I believe those are the only two possible situations that could happen at this point in the code

File: app/jobs/send_notification_job.rb [hunk 11]
```diff
@@ -27,54 +27,62 @@ class SendNotificationJob < CaseflowJob
...
+  #
+  # Response: JSON from VA Notify API
+  def send_to_va_notify(message)
+    event = NotificationEvent.find_by(event_type: message.template_name)
+    email_template_id = event.email_template_id
+    sms_template_id = event.sms_template_id
+
+    if FeatureToggle.enabled?(:va_notify_email)
+      VANotifyService.send_email_notifications(message.participant_id, message.appeal_id, email_template_id, status = "")
     end
```
Comment (jshields13) [inline app/jobs/send_notification_job.rb:82]:
```suggestion
    if FeatureToggle.enabled?(:va_notify_email)
      content = VANotifyService.send_email_notifications(message.participant_id, message.appeal_id, email_template_id, status = "")
      if !content.nil? && content != ""
          # update table
      end
    end
```

**Review by jshields13 — APPROVED** [review]

**Pull request merged**